### PR TITLE
Chapter 4 solutions

### PR DESCRIPTION
**Pull request by saatvikshah**
### Solutions for Chapter 4

Also found a possible incorrect test case too. I'll file an issue if you'll could confirm its indeed incorrect.
(Edit: The basic test cases did pass for me locally, so Im not sure why the GH action failed).

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter4.hs [hunk 1]
```diff
@@ -266,6 +273,10 @@ instance Functor Maybe where
     fmap f (Just a) = Just (f a)
     fmap _ x = x
 @
+
+A: Because "fmap _ x = x" which gets matched in the case of "Nothing", assumes that the return type and input type matches,
+which as we've seen above can be a different "Nothing" type. eg. in the `replicate 3`, case the type we are enforcing on the rhs of 
+that pattern match is "Maybe a", whereas it has been declared as `Maybe b` which corresponds to Maybe [a] for `replicate`.
```
Comment (vrom911) [inline src/Chapter4.hs:279]: 👏🏼

File: src/Chapter4.hs [hunk 2]
```diff
@@ -470,12 +487,16 @@ Applicatives can be found in many applications:
...
+-- Useful supplementary reading: https://adit.io/posts/2013-04-17-functors,_applicatives,_and_monads_in_pictures.html
+
 instance Applicative (Secret e) where
     pure :: a -> Secret e a
-    pure = error "pure Secret: Not implemented!"
+    pure = Reward
 
     (<*>) :: Secret e (a -> b) -> Secret e a -> Secret e b
-    (<*>) = error "(<*>) Secret: Not implemented!"
+    (Reward f) <*> x = fmap f x
```
Comment (vrom911) [inline src/Chapter4.hs:498]: Nice work on reusing `Functor` instance for `Secret`! :world_map:

File: src/Chapter4.hs [hunk 3]
```diff
@@ -489,6 +510,17 @@ Implement the 'Applicative' instance for our 'List' type.
   type.
 -}
 
+concatList :: List a -> List a -> List a
+concatList (Cons x xs) ys = Cons x (concatList xs ys)
```
Comment (vrom911) [inline src/Chapter4.hs:514]: Perfect!
Usually in Haskell, functions that add a few structures into are called `append*`, and contact is usually used for nested structured in order to flat them, but that is okay 🙂 

You can one more case to make your solution even more efficient:

```haskell
append xs Empty = xs
```

In that case, you won't need to walk through each element of the first list when the second one is empty and return the list instantly 🚤

File: src/Chapter4.hs [hunk 4]
```diff
@@ -628,8 +664,9 @@ Can you implement a monad version of AND, polymorphic over any monad?
 
 🕯 HINT: Use "(>>=)", "pure" and anonymous function
 -}
+
 andM :: (Monad m) => m Bool -> m Bool -> m Bool
-andM = error "andM: Not implemented!"
+andM xm ym = (ym >>= (\y -> pure (&& y))) <*> xm
```
Comment (vrom911) [inline src/Chapter4.hs:669]: The ordinary `&&` is lazy on the second argument. Meaning that if it meets `False` as the first argument, there is no need to calculate another arg. In `andM` we would like to keep the same behaviour, so the

```
Just False `andM` _
```

should be `Just False` :slightly_smiling_face: 


With `Applicative` you are not able to implement a lazier version of `andM`.

File: src/Chapter4.hs [hunk 5]
```diff
@@ -672,6 +709,22 @@ Specifically,
...
+
+-- Were the questions below supposed to be solved using the `fmap` above?
+
+reverseTree :: BTree a -> BTree a
+reverseTree (Node x xleft xright) = Node x xright xleft
+reverseTree Leaf = Leaf
+
+toList :: BTree a -> [a]
+toList (Node x xleft xright) = x : toList xleft ++ toList xright
+toList Leaf = []
```
Comment (vrom911) [inline src/Chapter4.hs:727]: Great Tree implementation 🌴 🎄 🎋

File: test/Test/Chapter4.hs [hunk 6]
```diff
@@ -56,7 +56,7 @@ chapter4advanced = describe "Chapter4Advanced" $
         it "Nothing - Nothing" $ andM Nothing Nothing `shouldBe` Nothing
         it "Nothing - Just" $ andM Nothing (Just True) `shouldBe` Nothing
         it "Just True - Nothing" $ andM (Just True) Nothing `shouldBe` Nothing
-        it "Just False - Nothing" $ andM (Just False) Nothing `shouldBe` Just False
+        it "Just False - Nothing" $ andM (Just False) Nothing `shouldBe` Just False -- is this test case correct?
```
Comment (vrom911) [inline test/Test/Chapter4.hs:59]: Yes, see the explanation above and let me know if that makes sense 🙂

**Review by vrom911 — APPROVED** [review]
Wow, you've finished the course :tada: Fantastic work!
Thank you so much for taking this course! I hope you had fun :man_dancing: 💃🏼 :microphone: 

We would really appreciate your [feedback on the course](https://docs.google.com/forms/d/e/1FAIpQLScBVhLxq5CgGnAfIGUE-fCoOUqeGkDY2HXzbT7KV2jjLOsmjQ/viewform)!

You can also ask any questions and we can discuss any topics you would like to dig into as well  :slightly_smiling_face:

**Comment by chshersh** [comment]
We can't merge solutions to our repository, so I'm closing this PR.
Feel free to open PRs to unsolved chapters in your fork of this repository 🍴

**Pull request closed without merging**